### DISCUSSION TITLE
Calculating the measure length

### DISCUSSION THREAD

**Discussion by cifkao** [salu133445/muspy#54] [General]
What is the formula for calculating the measure length in time steps? Could this information be added to the docs? https://muspy.readthedocs.io/en/latest/timing.html

A time signature specifies the number of beats per bar and how long a beat is. So the length of a bar in quarter notes should be something like `numerator * 4 // denominator`. But in time steps, it could be either `numerator * 4 * resolution // denominator`, or just `numerator * resolution`. Which of these is it? It's not clear because the docs sometimes say that `resolution` is the number of time steps per *beat* and sometimes per *quarter note*.

Moreover, there are compound meters like 6/8 and 9/8, where the beat duration is actually 3/8 and not 1/8. [This](https://github.com/craffel/pretty-midi/blob/5e3db4bfa6be0d6e87d7a9e8fcf5f4ed81e97a8d/pretty_midi/pretty_midi.py#L750-L755) is how `pretty_midi` handles them.

It's hard for me to wrap my head around all this. Should `TimeSignature` have another attribute specifying the number of time steps per beat or bar? This could be automatically computed, but perhaps it should be possible to set it explicitly in order to achieve some weird compound signatures.

**Comment by salu133445** [comment]
Well...they should all be _quarter note_ rather than _beat_. Did you remember where beats occur in the doc?

In general, I think beat is a rather vague concept in music. The issue about bpm vs qpm is somewhat similar to key vs key signature issue. MIDI does a really bad job on specifying symbolic features that do not affect playback. We could support this, but I doubt if there exist any files that contain such information without the need of inferring. Is there any field for such information in MusicXML other than time signatures?

**Reply by cifkao** [reply]
It says 'time steps per beat' in the `Music` docs in the table at the top: https://muspy.readthedocs.io/en/latest/classes/music.html

So the correct formula is `numerator * 4 * resolution // denominator`, at least for simple (non-compound) meters, right?
